src/broadcast/broadcast.rs

```diff
@@ -250,14 +279,66 @@ impl<N: NodeIdT> Broadcast<N> {
         // Save the proof for reconstructing the tree later.
         self.echos.insert(sender_id.clone(), p);
 
-        if self.ready_sent || self.count_echos(&hash) < self.netinfo.num_correct() {
+        // Upon receiving `N - 2f` `Echo`s with this root hash, multicast `CanDecode`
+        if !self.can_decode_sent && self.count_echos(&hash) >= self.netinfo.num_correct() - self.netinfo.num_faulty() {
+            return self.send_can_decode(&hash);
+        }
+
+        if self.ready_sent || self.count_echos(&hash) + self.count_echo_hashes(&hash) < self.netinfo.num_correct() {
             return self.compute_output(&hash);
         }
 
         // Upon receiving `N - f` `Echo`s with this root hash, multicast `Ready`.
         self.send_ready(&hash)
     }
 
+    fn handle_echo_hash(&mut self, sender_id: &N, hash: &Digest) -> Result<Step<N>> {
+        // If the sender has already sent `EchoHash`, ignore.
+        if let Some(old_hash) = self.echo_hashes.get(sender_id) {
+            if old_hash == hash {
+                warn!(
+                    "Node {:?} received EchoHash({:?}) multiple times from {:?}.",
+                    self.our_id(),
+                    hash,
+                    sender_id,
+                );
+                return Ok(Step::default());
+            } else {
+                return Ok(Fault::new(sender_id.clone(), FaultKind::MultipleEchoHashes).into());
+            }
+        }
+        // Save the hash for counting later.
+        self.echo_hashes.insert(sender_id.clone(), *hash);
+
+        if self.ready_sent || self.count_echos(&hash) + self.count_echo_hashes(&hash) < self.netinfo.num_correct() {
+            return self.compute_output(&hash);
+        }
+        // Upon receiving `N - f` `Echo`s with this root hash, multicast `Ready`.
+        self.send_ready(&hash)
+    }
+
+    /// Handles a received `CanDecode` message.
+    fn handle_can_decode(&mut self, sender_id: &N, hash: &Digest) -> Result<Step<N>> {
+        // If the sender has already sent `CanDecode`, ignore.
+        if let Some(old_hash) = self.can_decodes.get(sender_id) {
+            if old_hash == hash {
+                warn!(
+                    "Node {:?} received CanDecode({:?}) multiple times from {:?}.",
+                    self.our_id(),
+                    hash,
+                    sender_id,
+                );
+                return Ok(Step::default());
+            } else {
+                return Ok(Fault::new(sender_id.clone(), FaultKind::MultipleCanDecodes).into());
```

Comment:
I don't think we can attribute this fault to the sender; it could also be due to a faulty proposer: If the proposer sent one kind of matching `Value`s to one set of _N - f_ correct nodes, and another to another set of _N - f_ correct nodes, then poor `sender_id` could have received two sets of _N - f_ `Echo`s with different root hashes. I think in that case they actually have to send both `CanDecode`s, and we have to keep track of both! (Please double-check the proof; I think it only works that way.) So we'd need to make `can_decodes` a `BTreeSet<(N, Digest)>`, I guess?

Reply:
Makes sense. If a faulty proposer sends the conflicting messages `h0` and `h1` to the honest nodes such that neither set of honest nodes get `N-f` `Echo`s, then the protocol simply would stall. If one set gets `N-f` `Echos`, they propagate their `Ready`s and eventually, all honest nodes will be able to decode the majority `h_{i}`. The honest nodes have no way of knowing which `h_{i}`  eventually gets the majority votes (if there is a majority). Hence, they need to advertise their ability to decode both the `h_{i}`s to all the correct nodes to ensure that all the honest nodes can agree upon a value if possible.
In the first case(stalling) , does the protocol terminate somehow? We can't have timeouts here because of the asynchronous nature of the protocol I assume?

Reply:
You're exactly right, and the protocol can stall if the sender is malicious: if they don't send anything, it's guaranteed to stall, and if they send different messages to different groups of nodes, either all of them agree on one of those messages, or none of them terminates.

Inside Honey Badger, it's fine and expected if some Broadcast instances stall: A batch can be produced when at least ⅔ of them have terminated in the current epoch, which is guaranteed if at least ⅔ of nodes are correct.